library/alloc/src/raw_vec.rs

```diff
@@ -114,68 +121,95 @@ impl<T> RawVec<T, Global> {
-    // - 4 if elements are moderate-sized (<= 1 KiB).
-    // - 1 otherwise, to avoid wasting too much space for very short Vecs.
-    pub(crate) const MIN_NON_ZERO_CAP: usize = if mem::size_of::<T>() == 1 {
+impl RawVecInner<Global> {
+    #[must_use]
+    #[rustc_const_stable(feature = "raw_vec_internals_const", since = "1.81")]
+    const fn new<T>() -> Self {
+        Self::new_in(Global, core::mem::align_of::<T>())
+    }
+
+    #[cfg(not(any(no_global_oom_handling, test)))]
+    #[must_use]
+    #[inline]
+    fn with_capacity(capacity: usize, elem_layout: Layout) -> Self {
+        match Self::try_allocate_in(capacity, AllocInit::Uninitialized, Global, elem_layout) {
+            Ok(res) => res,
+            Err(err) => handle_error(err),
+        }
+    }
+}
+
+// Tiny Vecs are dumb. Skip to:
+// - 8 if the element size is 1, because any heap allocators is likely
+//   to round up a request of less than 8 bytes to at least 8 bytes.
+// - 4 if elements are moderate-sized (<= 1 KiB).
+// - 1 otherwise, to avoid wasting too much space for very short Vecs.
+const fn min_non_zero_cap(size: usize) -> usize {
+    if size == 1 {
         8
-    } else if mem::size_of::<T>() <= 1024 {
+    } else if size <= 1024 {
         4
     } else {
         1
-    };
+    }
+}
+
+impl<T, A: Allocator> RawVec<T, A> {
+    #[cfg(not(no_global_oom_handling))]
+    pub(crate) const MIN_NON_ZERO_CAP: usize = min_non_zero_cap(size_of::<T>());
 
     /// Like `new`, but parameterized over the choice of allocator for
     /// the returned `RawVec`.
+    #[inline]
+    #[rustc_const_stable(feature = "raw_vec_internals_const", since = "1.81")]
     pub const fn new_in(alloc: A) -> Self {
-        // `cap: 0` means "unallocated". zero-sized types are ignored.
-        Self { ptr: Unique::dangling(), cap: Cap::ZERO, alloc }
+        Self { inner: RawVecInner::new_in(alloc, align_of::<T>()), _marker: PhantomData }
     }
 
     /// Like `with_capacity`, but parameterized over the choice of
     /// allocator for the returned `RawVec`.
     #[cfg(not(no_global_oom_handling))]
     #[inline]
     pub fn with_capacity_in(capacity: usize, alloc: A) -> Self {
-        match Self::try_allocate_in(capacity, AllocInit::Uninitialized, alloc) {
-            Ok(res) => res,
-            Err(err) => handle_error(err),
+        Self {
+            inner: RawVecInner::with_capacity_in(capacity, alloc, T::LAYOUT),
+            _marker: PhantomData,
         }
     }
 
     /// Like `try_with_capacity`, but parameterized over the choice of
     /// allocator for the returned `RawVec`.
     #[inline]
     pub fn try_with_capacity_in(capacity: usize, alloc: A) -> Result<Self, TryReserveError> {
-        Self::try_allocate_in(capacity, AllocInit::Uninitialized, alloc)
+        match RawVecInner::try_with_capacity_in(capacity, alloc, T::LAYOUT) {
+            Ok(inner) => Ok(Self { inner, _marker: PhantomData }),
+            Err(e) => Err(e),
+        }
```

Comment:
`?` does not optimize away. Here's what we get for this function if we use it:
```
    bb0: { 
        StorageLive(_3);
        StorageLive(_4);
        _4 = RawVecInner::<A>::try_allocate_in(move _1, const AllocInit::Uninitialized, move _2, const <T as core::mem::SizedTypeProperties>::LAYOUT) -> [return: bb3, unwind continue];
    }   

    bb1: { 
        unreachable;
    }   

    bb2: { 
        return;
    }   

    bb3: { 
        StorageLive(_8);
        StorageLive(_9);
        StorageLive(_10);
        _8 = discriminant(_4);
        switchInt(move _8) -> [0: bb5, 1: bb4, otherwise: bb1];
    }   

    bb4: { 
        _10 = move ((_4 as Err).0: collections::TryReserveError);
        StorageLive(_11);
        _11 = Result::<Infallible, TryReserveError>::Err(_10);
        _3 = ControlFlow::<Result<Infallible, TryReserveError>, RawVecInner<A>>::Break(move _11);
        StorageDead(_11);
        StorageDead(_10);
        StorageDead(_9);
        StorageDead(_8);
        StorageDead(_4);
        _5 = move ((_3 as Break).0: core::result::Result<core::convert::Infallible, collections::TryReserveError>);
        _12 = ((_5 as Err).0: collections::TryReserveError);
        _0 = Result::<RawVec<T, A>, TryReserveError>::Err(_12);
        StorageDead(_3);
        goto -> bb2;
    }   

    bb5: { 
        _9 = move ((_4 as Ok).0: raw_vec::RawVecInner<A>);
        _3 = ControlFlow::<Result<Infallible, TryReserveError>, RawVecInner<A>>::Continue(_9);
        StorageDead(_10);
        StorageDead(_9);
        StorageDead(_8);
        StorageDead(_4);
        _6 = move ((_3 as Continue).0: raw_vec::RawVecInner<A>);
        StorageDead(_3);
        StorageLive(_7);
        _7 = RawVec::<T, A> { inner: _6, _marker: const ZeroSized: PhantomData<T> };
        _0 = Result::<RawVec<T, A>, TryReserveError>::Ok(move _7);
        StorageDead(_7);
        goto -> bb2;
    }
```

Reply:
And for completeness, the current impl produces:
```
    bb0: { 
        StorageLive(_3);
        _3 = RawVecInner::<A>::try_allocate_in(move _1, const AllocInit::Uninitialized, move _2, const <T as core::mem::SizedTypeProperties>::LAYOUT) -> [return: bb5, unwind continue];
    }   

    bb1: { 
        unreachable;
    }   

    bb2: { 
        _7 = move ((_3 as Err).0: collections::TryReserveError);
        _0 = Result::<RawVec<T, A>, TryReserveError>::Err(_7);
        goto -> bb4;
    }   

    bb3: {
        _5 = move ((_3 as Ok).0: raw_vec::RawVecInner<A>);
        StorageLive(_6);
        _6 = RawVec::<T, A> { inner: _5, _marker: const ZeroSized: PhantomData<T> };
        _0 = Result::<RawVec<T, A>, TryReserveError>::Ok(move _6);
        StorageDead(_6);
        goto -> bb4;
    }   

    bb4: {
        StorageDead(_3);
        return;
    }   

    bb5: { 
        _4 = discriminant(_3);
        switchInt(move _4) -> [0: bb3, 1: bb2, otherwise: bb1];
    }
```